### PR TITLE
new functions started

### PR DESCRIPTION
**Pull request by Samad198**

### CODE REVIEW

**Comment by beforan** [comment]
I think in principle this is pretty close to what we want but I have one concern, around using `singletonKeys` (cos I can't remember exactly what it's doing)

Maybe you can do some further testing (the project is already set up to use jest - see `iaa.test.js`) based on the following:

Imagine a set of intervals: `[1,5], [1,3], [2,5]`

All the min and max values in that set are `1,1,2,3,5,5` but the *distinct* values (what I think `singletonKeys` would get) are `1,2,3,5`.

Let's try and take the mean of each set of values:
- `1,1,2,3,5,5` -> `17/6` = `2.83`
- `1,2,3,5` -> `11/4` = `2.75`

They don't match, so it's obviously important that we include how many times we have the same value, so I think just summing `singletonKeys` is wrong. Instead we need each key, times the number of participants that selected it, and then the divisor needs to include the correct count as well.

This should only affect the midpoints average, since that's the one that's using the actual dataset - the maxima one, as described, uses linspace to sample across the range, rather than submitted values within the range, so that one should be fine.

Also the midpoints function at the moment only selects values which equal the min and max value of the whole dataset, but it needs all minimum values and maximum values from each individual interval.

So for our above interval set, the min values are `1,1,2` not just `1`, and the max values are `5,3,5` not just `5`.

File: js/src/iaa.js [hunk 1]
```diff
@@ -58,6 +58,39 @@ class IntervalAgreementApproach {
...
+    
+    return mean;
+  }
+
+  /**
+   * sum all minimum values and all maximum values then divide 
+   * by the count of all summed values (i.e. sum of the counts of min and max)
+   */
+  get mean_of_midpoints() {
+    const minValues = this.intervals.singletonKeys.filter(val=> val == Math.min(...this.intervals.singletonKeys))
```
Comment (beforan) [inline js/src/iaa.js]: This only selects values which equal the min and max value of the whole dataset, but it needs all minimum values and maximum values from each individual interval.

So for the interval set `[1,5], [1,3], [2,5]`, the min values are `1,1,2` not just `1`, and the max values are `5,3,5` not just `5`.

File: js/src/iaa.js [hunk 2]
```diff
@@ -58,6 +58,39 @@ class IntervalAgreementApproach {
...
+    const maxValues = this.intervals.singletonKeys.filter(val=> val == Math.max(...this.intervals.singletonKeys))
```
Comment (beforan) [inline js/src/iaa.js]: I don't think singletonKeys works for us here:

Imagine a set of intervals: `[1,5], [1,3], [2,5]`

All the min and max values in that set are `1,1,2,3,5,5` but the *distinct* values (what I think `singletonKeys` would get) are `1,2,3,5`.

Let's try and take the mean of each set of values:
- `1,1,2,3,5,5` -> `17/6` = `2.83`
- `1,2,3,5` -> `11/4` = `2.75`

They don't match, so it's obviously important that we include how many times we have the same value, so I think just summing `singletonKeys` is wrong. Instead we need each key, times the number of participants that selected it, and then the divisor needs to include the correct count as well.

**Review by beforan — CHANGES_REQUESTED** [review]

**Comment by Samad198** [comment]
I'm not sure if mean of maxima function is correct or not.
If there is a set of intervals: [1,5], [1,3], [2,5], should the return value be
(5+3+5)/3 which is 4.33? because it currently returns 3

**Review by beforan — APPROVED** [review]

**Pull request merged**